### DISCUSSION TITLE
plugin serve static files and load into custom panel

### DISCUSSION THREAD

**Discussion by wolflu05** [inventree/InvenTree#4020] [Q&A] [answered]
Hello, I'm currently continuing working on my plugin. It has a complex reactive form which I don't want to do by hand in vanilla js, so I decided to try out preact in combination with htm. However to use this, I would need to load static scripts which add something to the window object. Is there any way where I can place my scripts to serve them via `/static` and use them in custom panels? I know there is this `javascript_template` option in the panels array, but how do I add static files, because I already have my js code load at `javascript_template`.

**Comment by SchrodingersGat** [comment, marked as answer]
Unless @matmair has another suggestion, I think that this would be difficult in the current architecture, for a number of reasons.

First and foremost is how static files are served. They must be stored under the `STATIC_ROOT` directory (on your filesystem). The plugins do not have access to this directory, and exposing external directories to the fileserver is a huge security hole.

Secondly, defining how pathing to plugin files under `/static/` works, without shadowing required files, is problematic. 

Perhaps there is a way to make this work, but my suggestion would be:

**Template Files:** Include the files you want as part of your plugin template directory, and load them into the webpage

**External Files:** Nothing stopping you loading files from an external service

**Upload Media Files:** You can upload custom files to the `/media/` directory and access them via `/media/`

**Upload Static Files:** You can always place your own custom files within the `STATIC_ROOT` directory on your filesystem, but you need access to the InvenTree server to do this. And you need to do it manually.

**Reply by wolflu05** [reply]
Thank you @SchrodingersGat for your reply. If I try that:
> Template Files: Include the files you want as part of your plugin template directory, and load them into the webpage

They get returned as a html content type so they won't load as scripts. 

And to your point with external scripts, I'm not sure if it's the right way due to security concerns if companies are using my plugin.


And a furthermore my issue is how to load them into a panels html so that I can assume in my panel JavaScript_template they are already loaded and use functions they provide.

**Reply by matmair** [reply]
@wolflu05 if you use the AppMixin you can define a static dir as with any other django app - that way the files get colllected when running `colllectstatic`.

**Reply by wolflu05** [reply]
> @wolflu05 if you use the AppMixin you can define a static dir as with any other django app - that way the files get colllected when running `colllectstatic`.

Good idea, I got this to working by creating a `settings.py` file with the following contents: 
```py
from pathlib import Path

STATIC_ROOT = Path("./static").resolve()
```

Then I can serve the files via `/static/preact.js` however, I'm still struggling how to include them, so that they are loaded to the time the content from the panels `javascript` content is executed. Do you also have an idea how to solve that?

**Reply by wolflu05** [reply]
I have figured it out. This is my final solution, thank you all:

1. Add `AppMixin` to your plugin
2. Add your static files to a folder called `static/<your_plugin_name>` in your plugin's directory
3. Load the static files in your template like this:
   ```html
   {% load static %}
   <script src="{% static '<your_plugin_name>/preact.js' %}"></script>
   ```
4. Goto your `javascript_template` of your panel and assume that the js static files are already loaded. It will work. (I had caching issues in my previous test so that's why it didn't work)
5. Add this to your `MANIFEST.in` file in the root of your repo, where the `setup.py` lives.
   ```
   recursive-include path/to/your/static *
   ```
   And make sure, that you have set the `include_package_data=True` option in your `setup.py` file so that your static files are included in the pypi bundle .

> ℹ️ Don't put your static files directly in the static folder, because they may can interfere with the static files from inventree core.

**Reply by SchrodingersGat** [reply]
@matmair or @wolflu05  would you be able to add some notes to the documentation about this?

**Reply by matmair** [reply]
@wolflu05 I do not really understand why you need 2. It should work without it.

Regarding 3: Per Django convention scout should place all your plugins(apps) static files in a dir named like the plugin in the static dir. Adjust the static url in the template accordingly. You could have filename clashes else.

In generell I strongly discourage any changes to files provided from upstream/inventree - we might change around anything that is not defined as customizable in the docs at any time.

**Reply by matmair** [reply]
@SchrodingersGat using custom static files is a very advanced pattern, I would question if we want to support this as documenting it signals support from the core team.

**Reply by wolflu05** [reply]
> In generell I strongly discourage any changes to files provided from upstream/inventree > - we might change around anything that is not defined as customizable in the docs at any time.

What do I need to change in the inventree source to make this to work? It's working without an changes for me.

> using custom static files is a very advanced pattern, I would question if we want to support this as documenting it signals support from the core team.

I don't think this is an advanced pattern to have the need of static files. This can happen in every plugin that exposes custom panels or templates.

**Reply by matmair** [reply]
> 
> What do I need to change in the inventree source to make this to work? It's working without an changes for me.
> 
I think nothing but in step 2 above you touch settings.py -> which is part of core
> 
> I don't think this is an advanced pattern to have the need of static files. This can happen in every plugin that exposes custom panels or templates.

Agree to disagree; once novices start using the AppMixin we might as well start a Django help site

**Reply by wolflu05** [reply]
> I think nothing but in step 2 above you touch settings.py -> which is part of core

Is this inventree core logic? I thought this is Django logic. In my first try I didn't hat this settings.py created in my plugins, and it didn't worked, then I created it and it worked. Could be also working before, I'll test this again.

So what is the resume? @SchrodingersGat What do you think about the use of static files?

**Reply by matmair** [reply]
@wolflu05 the settings.py is an integral part of InvenTree's function, it won't work with variables missing there for example. Once you change that file automatic updates via installer or new image tag will not work.

**Reply by wolflu05** [reply]
You know that I‘m speaking of changing the settings.py in my plugins folder, not the inventree folder?

**Reply by matmair** [reply]
No, that was not clear in the text. I still do not understand why you need the `settings.py` in the plugins(?) subdir?

**Reply by wolflu05** [reply]
As I said, I tested onetime without, and it didn't work. Then I added it and it worked. But it could be also the case that I didn't restated the server (used hot reload) on my first try without the settings.py. I'll try it without again if I have time.

**Reply by wolflu05** [reply]
@matmair you're right. It works without. Edited my solution.

**Reply by matmair** [reply]
@wolflu05 thanks for checking. I was really confused seeing that I wrote that whole subsystem ;-)
